_posts/2020-12-21--kinesis-streams-part-2.md

```diff
@@ -0,0 +1,591 @@
+
+One thing to note here is that you can not use Lambda’s own retry and DLQ settings, because they are meant only for asynchronous Lambda invocations and, as we have learned, the _ESM_ invokes a Lambda function **synchronously**.
+
+
+![async](/img/kinesis/async.png){: .img.centered }
+
+
+Instead, you need to configure the on-failure destination and other error handling settings on the **event source mapping**, when defining your function trigger.
+
+
+![sync](/img/kinesis/sync.png){: .img.centered }
+
+
+
+
+### Why bad things happen to good streams
+
+Now we know how to deal with errors in the Lambda function itself. But how about system errors? Why do they happen in the first place and should we do something about them as well?
+
+Up until now, we have only been concentrating on what happens **after** the _ESM_ gets a batch of records from a stream. But we haven’t talked much about **how exactly the _ESM_ gets those batches**. The thing is, it depends on whether you are using a shared or a dedicated throughput consumer. To better understand what is happening behind the scenes, let’s compare how _Lambda service_ and the _ESM_ behave in both these scenarios.
+
+
+### Shared throughput
+
+As briefly mentioned before, the shared throughput consumers (also known as the **standard iterator**) use the HTTP to poll records from the stream. What this actually means in the case of Lambda is that the _ESM_ polls the shards behind the scenes, using the Kinesis **_GetRecords API_** calls.
+
+An important detail to note here is that no matter what your batch window or size is, or whether you have a parallelization factor configured or not, the _ESM_ will do the polling at the steady rate of exactly **one _GetRecords_ call per second** for each shard in your stream. Your Lambda function may of course be invoked more often than once per second per shard (if the configured batch size is reached earlier), but under the hood, the _ESM_ will use the records it has buffered with the previous _GetRecords_ calls. This also means that the records your Lambda function gets can be as much as **one second old** (aka, the record propagation latency is 1 second).
+
+The _GetRecords API_ comes with some important [limits](https://docs.aws.amazon.com/streams/latest/dev/service-sizes-and-limits.html) we haven’t discussed yet. First of all, there is a limit on **how much data** can a single _GetRecords_ call retrieve from a shard (either 10 000 records or 10 MB, whichever comes first). The good thing is, in the case of Lambda, the _ESM_ automatically makes sure this limit is never reached (remember, you can not have a batch bigger than 10 000 records?).
```

Comment:
Unnecessary question mark? "...you can not have a batch bigger than 10 000 records"